pkg/signature/kms/cliplugin/signer_test.go

```diff
@@ -0,0 +1,306 @@
+				}
+				if diff := cmp.Diff(common.ProtocolVersion, args[0]); diff != "" {
+					t.Errorf("unexpected protocol version (-want +got):\n%s", diff)
+				}
+				if stdinBytes, err := io.ReadAll(stdin); err != nil {
+					t.Errorf("reading stdin: %v", err)
+				} else if diff := cmp.Diff(testStdinBytes, stdinBytes); diff != "" {
+					t.Errorf("unexpected stdin bytes (-want +got):\n%s", diff)
+				}
+				osArgs := append([]string{name}, args...)
+				if pluginArgs, err := handler.GetPluginArgs(osArgs); err != nil {
+					t.Error(err)
+				} else if diff := cmp.Diff(testPluginArgs, pluginArgs); diff != "" {
+					t.Errorf("parsing plugin args (-want +got):\n%s", diff)
+				}
+				return testCmd{
+					output: tc.cmdOutputBytes,
+					err:    tc.cmdOutputErr,
+				}
+			}
+			// client with our mocked Cmd
+			testPluginClient := newPluginClient(testExecutable, testInitOptions, makeCmdFunc)
+			// invokePlugin
+			testContext := context.TODO()
+			testStdin := bytes.NewBuffer(testStdinBytes)
+			resp, err := testPluginClient.invokePlugin(testContext, testStdin, testMethodArgs)
+			// compare results
+			if respDiff := cmp.Diff(tc.resp, resp); respDiff != "" {
+				t.Errorf("unexpected resp (-want +got):\n%s", respDiff)
+			}
+			if errorDiff := cmp.Diff(tc.err, err, cmpopts.EquateErrors()); errorDiff != "" {
+				t.Errorf("unexpected error (-want +got):\n%s", errorDiff)
+			}
+			if err != nil && !strings.Contains(err.Error(), tc.errorMessageSubstring) {
+				t.Errorf("error content does not contain expecrted substring (-want +got): \n-%s\n+%s", tc.errorMessageSubstring, err.Error())
+			}
+		})
+	}
+}
+
+// testSignerVerifierImpl is a mock implementation that asserts that the
+// expected values are both sent and received through the encoding and decoding processes.
+type testSignerVerifierImpl struct {
+	// TODO: remove this embedding after all methods are implemented.
+	kms.SignerVerifier
+	t *testing.T
+}
+
+// DefaultAlgorithm accepts no arguments, but returns an expected value.
+func (s testSignerVerifierImpl) DefaultAlgorithm() string {
+	return testDefaultAlgorithm
+}
+
+// CreateKey checks the expected context deadline and algorithm, and returns the expected public key.
+func (s testSignerVerifierImpl) CreateKey(ctx context.Context, algorithm string) (crypto.PublicKey, error) {
+	if diff := cmp.Diff(testDefaultAlgorithm, algorithm); diff != "" {
+		s.t.Errorf("unexpected algorithm (-want +got):\n%s", diff)
+	}
+	ctxDeadline, ok := ctx.Deadline()
+	if !ok {
+		s.t.Errorf("expected a deadline")
+	}
+	if diff := cmp.Diff(testContextDeadline, ctxDeadline); diff != "" {
+		s.t.Errorf("unexpected context deadline (-want +got):\n%s", diff)
+	}
+	return testPublicKey, nil
+}
+
+// TestPluginClient tests each of PluginClient's methods for correct encoding and decoding between a simulated plugin program,
+// by mocking the Cmd function and using TestSignerVerifierImpl to both check and return expected values.
+func TestPluginClient(t *testing.T) {
```

Comment:
What is the test for? It looks like all calls are mocked so it's mainly testing that an interface is implemented. Am I missing something?

Reply:
It's to check that, for each method, all the arguments and return values can survive the JSON serialization and deserialization that happens through invokePlugin.